Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,186 @@ lemma iff_flat_and_lTensor_reflects_triviality :
+    exact le_antisymm triv_coh complex
+
+  -- Since `M` is faithfully flat, we need only to show that `H ⊗ M` is trivial.
+  suffices Subsingleton (H ⊗[R] M) from rTensor_reflects_triviality R M H
+  let e : H ⊗[R] M ≃ₗ[R] _ := TensorProduct.quotientTensorEquiv _ _
+  -- Note that `H ⊗ M` is isomorphic to `ker l12 ⊗ M ⧸ range ((range l12 ⊗ M) -> (ker l23 ⊗ M))`.
+  -- So the problem is reduced to proving surjectivity of `range l12 ⊗ M → ker l23 ⊗ M`.
+  rw [e.toEquiv.subsingleton_congr, Submodule.subsingleton_quotient_iff_eq_top,
+    LinearMap.range_eq_top]
+  intro x
+  induction x using TensorProduct.induction_on with
+  | zero => exact ⟨0, by simp⟩
+  -- let `x ⊗ m` be an element in `ker l23 ⊗ M`, then `x ⊗ m` is in the kernel of `l23 ⊗ 𝟙M`.
+  -- Since `N1 ⊗ M -l12 ⊗ M-> N2 ⊗ M -l23 ⊗ M-> N3 ⊗ M` is exact, we have that `x ⊗ m` is in
+  -- the range of `l12 ⊗ 𝟙M`, i.e. `x ⊗ m = (l12 ⊗ 𝟙M) y` for some `y ∈ N1 ⊗ M` as elements of
+  -- `N2 ⊗ M`. We need to prove that `x ⊗ m = (l12 ⊗ 𝟙M) y` still holds in `(ker l23) ⊗ M`.
+  -- This is okay because `M` is flat and `ker l23 -> N2` is injective.
+  | tmul x m =>
+    rcases x with ⟨x, (hx : l23 x = 0)⟩
+    have mem : x ⊗ₜ[R] m ∈ LinearMap.ker (l23.rTensor M) := by simp [hx]
+    rw [LinearMap.exact_iff.1 ex] at mem
+    obtain ⟨y, hy⟩ := mem
+
+    refine ⟨LinearMap.rTensor M (LinearMap.rangeRestrict _ ∘ₗ LinearMap.rangeRestrict l12) y,
+      Module.Flat.rTensor_preserves_injective_linearMap (LinearMap.ker l23).subtype
+      Subtype.val_injective ?_⟩
+    simp only [LinearMap.comp_codRestrict, LinearMap.rTensor_tmul, Submodule.coe_subtype, ← hy]
+    rw [← LinearMap.comp_apply]
+    erw [← LinearMap.rTensor_comp]
+    rw [← LinearMap.comp_apply, ← LinearMap.rTensor_comp]
+    rfl
+  | add x y hx hy =>
+    obtain ⟨x, rfl⟩ := hx; obtain ⟨y, rfl⟩ := hy
+    exact ⟨x + y, by simp⟩
+
+lemma lTensor_reflects_exact [fl : FaithfullyFlat R M]
+    (N1 N2 N3 : Type*)
+    [AddCommGroup N1] [Module R N1]
+    [AddCommGroup N2] [Module R N2]
+    [AddCommGroup N3] [Module R N3]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+    (ex : Function.Exact (l12.lTensor M) (l23.lTensor M)) :
+    Function.Exact l12 l23 :=
+  rTensor_reflects_exact R M _ _ <| ex.of_ladder_linearEquiv_of_exact
+    (e₁ := TensorProduct.comm _ _ _) (e₂ := TensorProduct.comm _ _ _)
+    (e₃ := TensorProduct.comm _ _ _) (by ext; rfl) (by ext; rfl)
+
+end arbitrary_universe
+
+lemma exact_iff_rTensor_exact [fl : FaithfullyFlat R M]
+    (N1 N2 N3 : Type max u v)
+    [AddCommGroup N1] [Module R N1]
+    [AddCommGroup N2] [Module R N2]
+    [AddCommGroup N3] [Module R N3]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3) :
+    Function.Exact l12 l23 ↔ Function.Exact (l12.rTensor M) (l23.rTensor M) :=
+  ⟨fun e => Module.Flat.iff_rTensor_exact.1 fl.toFlat e,
+    fun ex => rTensor_reflects_exact R M l12 l23 ex⟩
+
+lemma iff_exact_iff_rTensor_exact :
+    FaithfullyFlat R M ↔
+    (∀ (N1 N2 N3 : Type max u v)
```

Comment:
Make sure to use the same implicitness as in the other statements in the `∀` condition (unless you have a good reason to not do so).

Reply:
I have changed everything to implicit variable